File: update-core/src/main/java/dev/bee/kanjianki/updatecore/AutoUpdateStatusPolicy.java
package dev.bee.kanjianki.updatecore;

public final class AutoUpdateStatusPolicy {
    public static final String DEFAULT_LAST_RESULT = "No automatic update check has run yet.";

    private AutoUpdateStatusPolicy() {
    }

    public static StatusFields normalize(
            boolean enabled,
            long lastCheckAtMillis,
            String lastResult,
            String lastVersion,
            String pendingApkName,
            String pendingMessage
    ) {
        return new StatusFields(
                enabled,
                lastCheckAtMillis,
                text(lastResult),
                text(lastVersion),
                text(pendingApkName),
                text(pendingMessage)
        );
    }

    public static boolean hasPendingUpdate(String pendingApkName) {
        return !text(pendingApkName).isEmpty();
    }

    public static String text(String value) {
        return value == null ? "" : value;
    }

    public record StatusFields(
            boolean enabled,
            long lastCheckAtMillis,
            String lastResult,
            String lastVersion,
            String pendingApkName,
            String pendingMessage
    ) {
    }
}

File: update-core/src/main/kotlin/dev/bee/kanjianki/updatecore/AutoUpdateStatusPolicy.kt
package dev.bee.kanjianki.updatecore

object AutoUpdateStatusPolicy {
    const val DEFAULT_LAST_RESULT = "No automatic update check has run yet."

    @JvmStatic
    fun normalize(
        enabled: Boolean,
        lastCheckAtMillis: Long,
        lastResult: String?,
        lastVersion: String?,
        pendingApkName: String?,
        pendingMessage: String?,
    ): StatusFields {
        return StatusFields(
            enabled,
            lastCheckAtMillis,
            text(lastResult),
            text(lastVersion),
            text(pendingApkName),
            text(pendingMessage),
        )
    }

    @JvmStatic
    fun hasPendingUpdate(pendingApkName: String?): Boolean {
        return text(pendingApkName).isNotEmpty()
    }

    @JvmStatic
    fun text(value: String?): String {
        return value ?: ""
    }

    class StatusFields(
        private val enabled: Boolean,
        private val lastCheckAtMillis: Long,
        private val lastResult: String,
        private val lastVersion: String,
        private val pendingApkName: String,
        private val pendingMessage: String,
    ) {
        fun enabled(): Boolean = enabled
        fun lastCheckAtMillis(): Long = lastCheckAtMillis
        fun lastResult(): String = lastResult
        fun lastVersion(): String = lastVersion
        fun pendingApkName(): String = pendingApkName
        fun pendingMessage(): String = pendingMessage
    }
}
